SCAutolib/models/card.py

```diff
@@ -0,0 +1,229 @@
+        """
+        End of context manager for virtual smart card. If any exception was
+        raised in the current context, it would be logged as an error.
+
+        :param exp_type: Type of the exception
+        :param exp_value: Value for the exception
+        :param exp_traceback: Traceback of the exception
+        """
+        if exp_type is not None:
+            logger.error("Exception in virtual smart card context")
+            logger.error(format_exc())
+        self.remove()
+
+    @property
+    def softhsm2_conf(self):
+        return self._softhsm2_conf
+
+    @softhsm2_conf.setter
+    def softhsm2_conf(self, conf: SoftHSM2Conf):
+        assert conf.conf_path.exists(), "File doesn't exist"
+        self._softhsm2_conf = conf.conf_path
+
+    def insert(self):
+        """
+        Insert virtual smart card by starting the corresponding service.
+        """
+        cmd = ["systemctl", "start", self._service_name]
+        out = run(cmd, check=True)
+        time.sleep(2)  # to prevent error with fast restarting of the service
+        logger.info(f"Smart card {self._service_name} is inserted")
+        return out
+
+    def remove(self):
+        """
+        Remove the virtual card by stopping the service
+        """
+        cmd = ["systemctl", "stop", self._service_name]
+        out = run(cmd, check=True)
+        time.sleep(2)  # to prevent error with fast restarting of the service
+        logger.info(f"Smart card {self._service_name} is removed")
+        return out
+
+    def enroll(self):
+        """
+        Upload certificate and private key to the virtual smart card (upload to
+        NSS database) with pkcs11-tool.
+        """
+        cmd = ["pkcs11-tool", "--module", "libsofthsm2.so", "--slot-index",
+               '0', "-w", self._private_key, "-y", "privkey", "--label",
+               f"'{self.user.username}'", "-p", self.user.pin, "--set-id", "0",
+               "-d", "0"]
+        run(cmd, env={"SOFTHSM2_CONF": self._softhsm2_conf})
+        logger.debug(
+            f"User key {self._private_key} is added to virtual smart card")
+
+        cmd = ['pkcs11-tool', '--module', 'libsofthsm2.so', '--slot-index', "0",
+               '-w', self._cert, '-y', 'cert', '-p', self.user.pin,
+               '--label', f"'{self.user.username}'", '--set-id', "0", '-d', "0"]
+        run(cmd, env={"SOFTHSM2_CONF": self._softhsm2_conf})
+        logger.debug(
+            f"User certificate {self._cert} is added to virtual smart card")
+
+    def create(self):
+        """
+        Creates SoftHSM2 token and systemd service for virtual smart card.
+        Directory for NSS database is created in this method as separate DB is
+        required for each virtual card.
+        """
+
+        assert self._softhsm2_conf.exists(),\
+            "Can proceed, SoftHSM2 conf doesn't exist"
+        Path(f"{self.user.card_dir}/tokens").mkdir()
+
+        p11lib = "/usr/lib64/pkcs11/libsofthsm2.so"
```

Comment:
Is there a reason for this path to be hardcoded here? Could it be defined in the class with other properties and called with self.p11lib? It's kind of hidden here.

Reply:
this library shouldn't be one another location, so there is no reason to provide an opportunity to change it as it can lead to more errors because the user might don't know what to do with this